Refactor(Purchase Order Registry)

lomamech opened the pull request:
Refactor(Purchase Order Registry) 
- Integrate grid save State
- Add dropdaown for column configuration, Download report on CSV and PDF 
- Implement report Purchase Order Handlebars

test(Purchase Order Registry) 
- Implement E2E test for check Purchase order registry 
- Fix some property and Identification
 - Remove old E2E test

closes #1746

lomamech committed:
Refactor(Purchase Order Registry)
- Integrate grid save State
- Add dropdaown for column configuration, Download report on CSV and PDF
- Implement report Purchase Order Handlebars

lomamech committed:
test(Purchase Order Registry)
- Implement E2E test for check Purchase order registry
- Fix some property and Identification
- Remove old E2E test

jniles commented on client/src/modules/purchases/list/list.js:
```diff
@@ -89,37 +91,34 @@ function PurchaseListController ($translate, PurchaseOrder, Notify, uiGridConsta
+    enableSorting     : true,
+    enableColumnMenus : false,
     flatEntityAccess  : true,
+    fastWatch         : true,
```
:+1:

jniles commented on client/src/modules/purchases/modals/search.modal.js:
```diff
@@ -1,24 +1,55 @@
 
-function SearchPurchaseOrderModalController(Suppliers, Notify, Instance, SearchFilterFormat) {
+/**
+ * @class PatientRegistryModalController
```
This should probably talk about the `SearchPurchaseOrderModalController()` instead.

jniles commented on client/src/modules/purchases/modals/search.modal.js:
```diff
@@ -27,27 +58,52 @@ function SearchPurchaseOrderModalController(Suppliers, Notify, Instance, SearchF
-      dateTo   : new Date(),
-    };
+  // custom filter user_id - assign the value to the params object
+  vm.onSelectUser = function onSelectUser(user) {
```
:+1:  Nice comments!

jniles commented on server/controllers/finance/purchases.js:
```diff
@@ -288,7 +289,20 @@ function search(req, res, next) {
+  // ensure epected options are parsed appropriately as binary
+  db.convert(options, ['supplier_uuid']);
+
+  const filters = new FilterParser(options, { tableAlias : 'p', autoParseStatements : false  });
```
:+1:

jniles commented on server/controllers/finance/reports/purchases/report.handlebars:
```diff
@@ -0,0 +1,83 @@
+              <td class="text-right">{{currency cost currency_id}}</td>
+              <td>
+                <strong>
+                  {{#if is_confirmed}}
```
Wow, this is a really confusing series of logic.  Can you move this into JS?  Something like:
```js
row.isPartiallyRecieved = (row.is_confirmed && !row.is_canceled && row.is_received && row.is_partially_received);
```
and then you can just 
```hbs
<strong>
  {{#if row.isPartiallyReceived }} {{translate 'PURCHASES.STATUS.PARTIALLY_RECIEVED' }} {{/if}}
  {{! etc ...}}
</strong>
```

lomamech committed:
test(Purchase Order)
- reset number of row from test "Change the status of PO and find 2
  purchases"

jniles commented on client/src/modules/purchases/list/list.html:
```diff
@@ -1,40 +1,92 @@
-      ui-grid-grouping>
+      ui-grid-move-columns
+      ui-grid-save-state
+      ui-grid-selection>
```
You never actually use `ui-grid-selection`.  It can be removed.

jniles commented on client/src/modules/purchases/modals/search.modal.js:
```diff
@@ -1,24 +1,55 @@
+  ];
+
+  // assign already defined custom params to searchQueries object
+  vm.params = util.maskObjectFromKeys(params, searchQueryOptions);
```
:+1:

jniles commented on client/src/modules/purchases/modals/search.tmpl.html:
```diff
@@ -1,88 +1,111 @@
   novalidate>
 
-  <div class="modal-header">
+  <div class="modal-header" style="box-shadow : 0px 2px 4px -4px black; z-index:1000">
```
You can just use `class="search-modal-header" and remove hard coded styles.

jniles commented on client/src/modules/purchases/modals/search.tmpl.html:
```diff
@@ -1,88 +1,111 @@
+              <strong translate>FORM.LABELS.STATUS</strong>
+            </p>
+
+            <label class="checkbox-inline">
```
Are you sure these should be checkboxes?  It looks like they do not work as expected.

For example, in the test dataset, search of "confirmed" yields 1 result.  Searching for both "confirmed" and "Received In Stock" should show _at least_ one, but nothing matches.
![confirmedandreceivedinstock](https://user-images.githubusercontent.com/896472/28961066-d9c96d8a-78f8-11e7-94fd-6e4721d6543d.png)
_Fig 1: Confirmed and Received Selection_

jniles commented on server/controllers/finance/purchases.js:
```diff
@@ -288,7 +289,20 @@ function search(req, res, next) {
+  filters.dateFrom('custion_period_start', 'date');
+  filters.dateTo('custom_period_end', 'date');
+  filters.equals('user_id');
+  filters.equals('is_confirmed');
```
Since you are using checkboxes, you would expect this to make an SQL like:
```sql
is_confirmed = ? OR is_received = ? OR is_canceled = ?
```
That way at least one matches when you search for both being checked.

jniles commented on server/controllers/finance/purchases.js:
```diff
@@ -302,16 +316,12 @@ function find(options) {
   const referenceStatement = `CONCAT_WS('.', '${identifiers.PURCHASE_ORDER.key}', pr.abbr, p.reference) = ?`;
   filters.custom('reference', referenceStatement);
 
-  filters.setOrder('ORDER BY p.date DESC');
```
You should probably keep the `setOrder()` around.  We want to make sure only recent records are displayed at the top.

jniles commented on client/src/modules/purchases/modals/search.modal.js:
```diff
@@ -1,24 +1,55 @@
+ * search functionality on the patient registry page.  Filters that are already
+ * applied to the grid can be passed in via the params inject.
+ */
+function SearchPurchaseOrderModalController(ModalInstance, params, Suppliers, bhConstants, moment, Store, util, Periods, Notify) {
```
Do you use `moment`?  Please remove any unused imports.

jniles commented on test/end-to-end/purchases/registry.spec.js:
```diff
@@ -0,0 +1,13 @@
+describe('Purchase Order Registry', function () {
+  'use strict';
+
+  before(() => helpers.navigate('#/purchases/list'));
```
It would be nice to make this URL `/purchases`.  That way it would match the other registries:
 1. `/patients` -> patient registry
 2. `/payments` -> cash payments registry
 3. `/invoices` -> invoice registry.

jniles commented on client/src/modules/purchases/list/list.html:
```diff
@@ -1,40 +1,92 @@
+          ng-click="PurchaseListCtrl.search()"
+          data-method="search"
+          class="btn btn-default">
+          <span class="fa fa-search"></span> <span translate>FORM.BUTTONS.SEARCH</span>
```
If you put `class="hidden-xs"` on the FORM.BUTTONS.SEARCH element, it will be hidden on small screens.  That makes a much nicer user experience :+1:

jniles reviewed (changes requested):
@lomamech, this is a really nice refactor.  It's beginning to look like a standard registry!

Some comments:
 1. The `is_confirmed`, `is_received`, and `is_canceled` do not behave correctly.  When combined, they will never match anything.  The solution is to write a custom filter (`filters.custom()` that will allow these to be matched with an OR.
 2. Please remove all unused imports (`require()` and `$injects`).
 3. The `amount` column does not respect sorting.  You need to put `type: 'number'` on the column `ui-grid`.
 4. Instead of using `Document`, it would be nice if this looked like the other registries.  For example, you could simply link the receipt from the `Reference` column, like the Patients Registry does.

After these edits are made, this will be completely up to date with 2.x standards.

lomamech committed:
refactor(Purchase Order Registry)
- Remove unused import
- Restaure filter setOrder
- Refactor purchases/report.handlebars use the logic in Js instead in html
- Use button radio instead checkbox for the research of Purchase Order
  Status
- Update the URL use #/purchases instead #/purchases/list

lomamech committed:
test(Purchase order): update result of test E2E

jniles committed:
refactor(purchases): minor cleanup of unused vars

This commit cleans up the purchase list code to remove unused variables and make sure error cases are appropriately reported to the grid.
